flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/PrintUtils.java

```diff
@@ -174,27 +209,264 @@ public static void printAsTableauForm(
+            Row row, ResolvedSchema resolvedSchema, ZoneId sessionTimeZone) {
+        return rowToString(row, NULL_COLUMN, false, resolvedSchema, sessionTimeZone);
     }
 
-    public static String[] rowToString(Row row, String nullColumn, boolean printRowKind) {
+    public static String[] rowToString(
+            Row row,
+            String nullColumn,
+            boolean printRowKind,
+            ResolvedSchema resolvedSchema,
+            ZoneId sessionTimeZone) {
         final int len = printRowKind ? row.getArity() + 1 : row.getArity();
         final List<String> fields = new ArrayList<>(len);
         if (printRowKind) {
             fields.add(row.getKind().shortString());
         }
         for (int i = 0; i < row.getArity(); i++) {
             final Object field = row.getField(i);
+            final LogicalType fieldType =
+                    resolvedSchema.getColumnDataTypes().get(i).getLogicalType();
             if (field == null) {
                 fields.add(nullColumn);
             } else {
-                fields.add(StringUtils.arrayAwareToString(field));
+                fields.add(
+                        StringUtils.arrayAwareToString(
+                                formattedTimestamp(field, fieldType, sessionTimeZone)));
             }
         }
         return fields.toArray(new String[0]);
     }
 
+    /**
+     * Normalizes field that contains TIMESTAMP and TIMESTAMP_LTZ type data.
+     *
+     * <p>This method also supports nested type ARRAY, ROW, MAP.
+     */
+    private static Object formattedTimestamp(
+            Object field, LogicalType fieldType, ZoneId sessionTimeZone) {
+        final LogicalTypeRoot typeRoot = fieldType.getTypeRoot();
+        if (field == null) {
+            return "null";
+        }
+        switch (typeRoot) {
+            case TIMESTAMP_WITHOUT_TIME_ZONE:
+            case TIMESTAMP_WITH_LOCAL_TIME_ZONE:
+                return formatTimestampData(field, fieldType, sessionTimeZone);
+            case ARRAY:
+                LogicalType elementType = ((ArrayType) fieldType).getElementType();
+                if (field instanceof List) {
+                    List<?> array = (List<?>) field;
+                    Object[] formattedArray = new Object[array.size()];
+                    for (int i = 0; i < array.size(); i++) {
+                        formattedArray[i] =
+                                formattedTimestamp(array.get(i), elementType, sessionTimeZone);
+                    }
+                    return formattedArray;
+                } else if (field instanceof ArrayData) {
+                    ArrayData array = (ArrayData) field;
+                    Object[] formattedArray = new Object[array.size()];
+                    for (int i = 0; i < array.size(); i++) {
+                        formattedArray[i] =
+                                formattedTimestamp(
+                                        array.getTimestamp(i, getPrecision(elementType)),
```

Comment:
This will throw exception if it's not an array of timestamp. I think we can ignore `ArrayData` for now.